Cython/Utility/ObjectHandling.c

```diff
@@ -282,35 +282,46 @@ static CYTHON_INLINE PyObject *__Pyx_PyObject_GetItem(PyObject *obj, PyObject* k
 // //@requires: GetItemInt - added in IndexNode as it uses templating.
 
 #if CYTHON_USE_TYPE_SLOTS
-static PyObject *__Pyx_PyObject_GetIndex(PyObject *obj, PyObject* index) {
-    PyObject *runerr;
-    Py_ssize_t key_value;
-    PySequenceMethods *m = Py_TYPE(obj)->tp_as_sequence;
-    if (unlikely(!(m && m->sq_item))) {
-        PyErr_Format(PyExc_TypeError, "'%.200s' object is not subscriptable", Py_TYPE(obj)->tp_name);
-        return NULL;
-    }
-
-    key_value = __Pyx_PyIndex_AsSsize_t(index);
-    if (likely(key_value != -1 || !(runerr = PyErr_Occurred()))) {
-        return __Pyx_GetItemInt_Fast(obj, key_value, 0, 1, 1);
+static PyObject *__Pyx_PyObject_GetItem(PyObject *obj, PyObject* key) {
+    PyTypeObject* tp = Py_TYPE(obj);
+    PyMappingMethods *mm = tp->tp_as_mapping;
+    if (likely(mm && mm->mp_subscript)) {
+        return mm->mp_subscript(obj, key);
+    }
+
+    PySequenceMethods *sm = tp->tp_as_sequence;
+    if (likely(sm && sm->sq_item)) {
+        PyObject *runerr;
+        Py_ssize_t key_value;
+        key_value = __Pyx_PyIndex_AsSsize_t(key);
+        if (likely(key_value != -1 || !(runerr = PyErr_Occurred()))) {
+            return __Pyx_GetItemInt_Fast(obj, key_value, 0, 1, 1);
+        }
+        if (PyErr_GivenExceptionMatches(runerr, PyExc_OverflowError)) {
+            PyErr_Clear();
+            PyErr_Format(PyExc_IndexError, "cannot fit '%.200s' into an index-sized integer", Py_TYPE(key)->tp_name);
+        }
+        return NULL; 
+    }
+    
+#if PY_VERSION_HEX >= 0x03070000
+    if (likely(PyType_Check(obj))) {
+        PyObject *meth;
+        _Py_IDENTIFIER(__class_getitem__);
+        if (_PyObject_LookupAttrId(obj, &PyId___class_getitem__, &meth) < 0) {
```

Comment:
I'd probably change this to `PyObject_GetAttrString(PYIDENT("__class_getitem__"))` which uses the Cython mechanisms for defining identifier names (it might have to be `PyObject_GetAttr`... not 100% sure).

Then I don't actually see the need to restrict to to Python 3.7+ - it's only an attribute lookup+function call so it should be possible for Cython to support it on any version of Python.

Reply:
`__Pyx_PyObject_GetAttrStr(obj, PYIDENT("__class_getitem__"))`, with an `//@requires: ObjectHandling.c::PyObjectGetAttrStr` at the top of the code section.

We're not using CPython's `_Py_IDENTIFIER` API in Cython.

Reply:
@da-woods I guess it depends on what the general strategy is for handling backwards compatibility. If I ran

```
class A:
  def __class_getitem__(cls, item):
    return item

print(A[2])
```

on Python 3.6 I should get a TypeError which I would no longer get without the python3.7 guard.

@scoder will update to use `__Pyx_PyObject_GetAttrStr` as suggested.